#Problem 1(a).  Unionized workers.

These problems use the 1985 Current Population Survey data, ```cps.csv```, from last week.

Create a pandas data frame to hold the CPS data.  Write a function that computes the average hourly wage of a unionized worker by using control flow to iterate through each row of the data frame.

In [2]:
import pandas as pd
import numpy as np

In [3]:
cps=pd.read_csv("/Users/becky/Desktop/cps.csv")
print(cps.iloc[:5,:])

   wage  educ race sex hispanic south  married  exper  union  age    sector
0   9.0    10    W   M       NH    NS  Married     27    Not   43     const
1   5.5    12    W   M       NH    NS  Married     20    Not   38     sales
2   3.8    12    W   F       NH    NS   Single      4    Not   22     sales
3  10.5    12    W   F       NH    NS  Married     29    Not   47  clerical
4  15.0    12    W   M       NH    NS  Married     40  Union   58     const


In [4]:
cps.shape[0]

534

In [5]:

#%%timeit -r1 -n1
def ave_hourly_wage_unionized(cps): #return true if unionezed
    unionized=0
    totalwage=0
    for ii in range(len(cps)):#for current row in cps.shape[0]
        row=cps.iloc[ii,:]#use iloc cuz referring to numeric indices and not loc for names
        if row['union']=='Union':
            unionized+=1
            totalwage=totalwage+row['wage']
    averagewage = totalwage/unionized
    return(averagewage)



In [6]:

cps2=ave_hourly_wage_unionized(cps)
cps2

10.798125000000001

#Problem 1(b).  More unionized workers.

Write a function that takes a row of a data frame as input.  Your function should return np.nan if a worker is not unionized, and the worker's hourly wage if the worker is unionized.  Apply this function to your data frame, then use the resulting series to compute the average hourly wage of a unionized worker.

In [8]:
def hourly(row):
    if row['union'] == "Union": #Union is a string
        return(row['wage'])
    else:
        return(np.nan)

def get_union_wages_2b(cps):    
    cps['unionWage']=cps.apply(hourly,axis=1)
    average_union2=np.mean(cps['unionWage'])
    return average_union2


get_union_wages_2b(cps)

10.798125

# Problem 1(c).  Selecting and analyzing unionized workers.

Write code that selects the unionized workers, then computes the average hourly wage of the resulting subset.

In [15]:
def get_union_wages_2c(cps):
    return cps[cps.union == 'Union'].wage.mean()
print('Union workers mean wage: ', cps[cps.union == 'Union'].wage.mean())


Union workers mean wage:  10.798125


In [10]:
%timeit ave_hourly_wage_unionized(cps)

119 ms ± 6.37 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [13]:
%timeit get_union_wages_2b(cps)

8.79 ms ± 95.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [16]:
%timeit get_union_wages_2c(cps)

779 µs ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


#Problem 1(d).  Timing.

Use the ```%time``` or ```%timeit``` magic commands to compare your code from 1(a), 1(b), and 1(c).  Create a markdown cell which explains the results of your comparison, including which is fastest and which is slowest.  Can you explain what makes the fastest code fastest?

1a: 119 ms ± 6.37 ms per loop (mean ± std. dev. of 7 runs, 10 loops each).  This is the slowest because it has a for loop so it has to go thru a lot more iterations

1b: 8.79 ms ± 95.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

1c: 779 µs ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each). This is the fastest because it's the most direct